br/pkg/checkpoint/storage.go

```diff
@@ -0,0 +1,333 @@
+			continue
+		}
+		if !bytes.Equal(uuid, lastUUID) {
+			if !lastUUIDInvalid && len(rowData) > 0 {
+				retData = append(retData, rowData)
+			}
+			rowData = make([]byte, 0)
+			lastUUIDInvalid = false
+			nextSegmentID = 0
+			lastUUID = uuid
+		}
+
+		if lastUUIDInvalid {
+			continue
+		}
+
+		if nextSegmentID != segment_id {
+			lastUUIDInvalid = true
+			continue
+		}
+
+		rowData = append(rowData, data...)
+		nextSegmentID += 1
+	}
+	if !lastUUIDInvalid && len(rowData) > 0 {
+		retData = append(retData, rowData)
+	}
+	return retData, nil
+}
+
+func selectCheckpointData[K KeyType, V ValueType](
+	ctx context.Context,
+	execCtx sqlexec.RestrictedSQLExecutor,
+	dbName string,
+	fn func(groupKey K, value V),
+) (time.Duration, error) {
+	// records the total time cost in the past executions
+	var pastDureTime time.Duration = 0
+	checkpointDatas, err := mergeSelectCheckpoint(ctx, execCtx, dbName, checkpointDataTableName)
+	if err != nil {
+		return pastDureTime, errors.Trace(err)
+	}
+	for _, content := range checkpointDatas {
+		if err := parseCheckpointData(content, &pastDureTime, nil, fn); err != nil {
+			return pastDureTime, errors.Trace(err)
+		}
+	}
+	return pastDureTime, nil
+}
+
+func selectCheckpointChecksum(
+	ctx context.Context,
+	execCtx sqlexec.RestrictedSQLExecutor,
+	dbName string,
+) (map[int64]*ChecksumItem, time.Duration, error) {
+	var pastDureTime time.Duration = 0
+	checkpointChecksum := make(map[int64]*ChecksumItem)
+	checkpointChecksums, err := mergeSelectCheckpoint(ctx, execCtx, dbName, checkpointChecksumTableName)
+	if err != nil {
+		return checkpointChecksum, pastDureTime, errors.Trace(err)
+	}
+	for _, content := range checkpointChecksums {
+		if err := parseCheckpointChecksum(content, checkpointChecksum, &pastDureTime); err != nil {
+			return checkpointChecksum, pastDureTime, errors.Trace(err)
+		}
+	}
+	return checkpointChecksum, pastDureTime, nil
+}
+
+func initCheckpointTable(ctx context.Context, se glue.Session, dbName string, checkpointTableNames []string) error {
+	if err := se.ExecuteInternal(ctx, "CREATE DATABASE IF NOT EXISTS %n;", dbName); err != nil {
+		return errors.Trace(err)
+	}
+	for _, tableName := range checkpointTableNames {
+		if err := se.ExecuteInternal(ctx, createCheckpointTable, dbName, tableName); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func insertCheckpointMeta[T any](ctx context.Context, se glue.Session, dbName string, tableName string, meta *T) error {
+	data, err := json.Marshal(meta)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := se.ExecuteInternal(ctx, createCheckpointMetaTable, dbName, tableName); err != nil {
+		return errors.Trace(err)
+	}
+	err = chunkInsertCheckpointData(data, func(segmentId uint64, chunk []byte) error {
+		err := se.ExecuteInternal(ctx, insertCheckpointMetaSQLTemplate, dbName, tableName, segmentId, chunk)
+		return errors.Trace(err)
+	})
+	return errors.Trace(err)
+}
+
+func selectCheckpointMeta[T any](
```

Comment:
It seems `meta` can be typed `any` and then we can get rid of the type parameter `T`.

Reply:
done